### ISSUE TITLE
user with EDIT access can undo PUBLIC access but cannot re-institute it. 

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#2080]
@dtarb @pkdash @hyi 
This quandary arose when testing new code. 

1. User Alice creates a resource and makes it `public`, and gives `edit` access to User Bob. 
2. User Bob deletes a file from the resource.. 
3. This invalidates `public` access because the file was required. The resource is set to `private`.
4. Bob adds back another version of the file, making the resource possible to make public.  
5. **Bob cannot set the resource to `public`, because Bob is not an `owner`.** 

Worse, this is the way we expect Bob to work. So `edit` users can't safely modify `public` resources.

**Comment by dtarb** [comment]
I think that we have to accept this as a limitation.  We could implement warnings, but I propose we do not at this stage change the functionality.

Here is what I suggest
- Implement a warning that if the last file is being deleted that the resource will revert to private and an owner will then need to set it to public.  Indicate in the warning that if the file is to be replaced, first adding the replacement file, then deleting the old file will retain the public setting

Rationale
- If Bob inadvertently ends up making the resource public, it is not that much trouble for them to contact the owner to have them reset it to public.  Owners and editors are supposed to be people that trust each other so working together.
- The work around above is functional
- We do want a resource that is required to have content, without any content to be private.

**Comment by alvacouch** [comment]
@dtarb The way we have done things, it is impractical to implement a pre-check. We won't know that the resource reverts to private until we delete the file and then run the validator. We don't have the concept of a pre-action validation. 

There are other possible solutions, including recording the "intent to make public" (which is different from the actual public flag). If one sets 'make_public_if_possible', then when Bob uploads a file that passes, the system takes over and acts on prior instructions of an owner to make something public.

**Comment by pkdash** [comment]
> Indicate in the warning that if the file is to be replaced, first adding the replacement file, then deleting the old file will retain the public setting

In some of the specific resource types, it won't be possible to add a replacement file before deleting the existing file. For example, if a time series resource already contains a SQLite file, it won't be possible to add another SQLite file.

**Comment by alvacouch** [comment]
The problem is that -- regardless of what we do to pre-validate -- we have to do something different for **every** resource type.  Lots of work. Hence "impractical".

**Comment by pkdash** [comment]
I am trying to get an understanding of the requirements for a file pre delete check function:

If we had a pre-delete-file-check() which says deleting this file will set the resource to private what we (system) suppose to do with that information?

**Comment by dtarb** [comment]
We already have a message "Confirm files deletion".  Lets just change this to stronger wording.

"This file will be permanently deleted.  Consider saving a copy if it is important to you.  **If this is the last file in the resource and it is public, the sharing status will revert to private.**  If you are not the owner of this resource, then an owner will need to reset this to public after a new file has been added.  If you want to replace this file, add the new file first then delete the old one, so that sharing status does not change."

The same problem occurs with deleting the last keyword.  We could add additional explanation to that message.

"Note: Deleting all keywords will set the resource sharing status to private.  If you are not the owner of this resource, then an owner will need to reset this to public after a new keyword has been added.  To avoid changing the sharing status add new keywords before deleting existing ones."

Regarding the concerns @pkdash raised about some resources only having one file, deleting the one and only file is essentially deleting the resource.  A new file added to the same container is essentially a new resource, so I am OK with it becoming private until the owner makes it public.  I think this is really a social issue for editors and owners and I do not see this as a major problem.  I see it as a relatively low priority minor inconvenience, and I am not sure pre and post change checks and balances are going to make it easier for a user, even if we do take the time, which appears considerable, to work it out.

**Comment by dtarb** [comment]
Assigned to @pkdash to change messages per above.  No change to functionality or policy at this time.

**Comment by alvacouch** [comment]
@dtarb @pkdash The "last file" condition is not the only condition under which a resource will become private. A resource with a set of shapefiles will not pass tests for public if some of the files are missing. 

This becomes more serious when people are "replacing" files. At present, this is done by deleting a file and uploading another. Because of the way validations are written, this turns off public if the resource without the file does not pass inspection!

**Comment by pkdash** [comment]
@alvacouch Agreed. File delete and file upload are two separate actions and not part of single transaction. The issue is that under the current implementation the system sets the public to private on file delete but it lacks the capability to set private to public when the resource has the required content files. However, If we keep provenance of resource flag (may be just a resource level bool type attribute ```private_set_by_system```) , then the system should be able to set private to public when the file gets added if that's what we want.

**Comment by dtarb** [comment]
While I have been reluctant to add additional complexity, I think some of the concerns raised here could be solved with a "public pending" sharing attribute.  This would be set in cases where public is intended, but where content (data files or metadata) does not pass validation rules.  Public pending resources would behave exactly as private, but instead of being labeled private would be labeled public pending on the landing page, and as soon as the content passes validation rules, they become public.  If content, through editing, results in failing validation, they become public pending.  Public pending would not be an attribute that an owner sets (remember, it is only owners that can set these sharing attributes).  Rather an owner would set public, and then the state public or public pending would be the result of rule validation.  Public pending would make the intent to be public transparent and automatic for editors, rather than reverting to private.

**Comment by hyi** [comment]
@dtarb I like this "public pending" approach which makes a lot of sense to me, and does not appear to me to involve a lot of work and complexity to implement. What do you think @pkdash @alvacouch ?

**Comment by dtarb** [comment]
Reviewing this as part of the old issue clean up, I would like to implement the warning message approach.  @pkdash can we work together on this once your workload on model aggregations subsides.

**Comment by dtarb** [comment]
Message to add to confirm files deletion box if the file is the last one being deleted and is going to revert sharing status to private.  “Deleting this file will set the resource sharing status to private.”

**Comment by eclark-cuahsi** [comment]
adding @dtarb as this pertains to Access and Control.

**Comment by dtarb** [comment]
Verified - closing.
